src/network/resolver.rs

```diff
@@ -0,0 +1,89 @@
+use std::{
+    collections::HashMap,
+    net::Ipv4Addr,
+    sync::{Arc, Mutex},
+    thread::{Builder, JoinHandle},
+};
+use tokio::{
+    runtime::Runtime,
+    sync::mpsc::{self, Sender},
+};
+use trust_dns_resolver::AsyncResolver;
+
+pub type Result<T> = std::result::Result<T, failure::Error>;
+pub type IpTable = Arc<Mutex<HashMap<Ipv4Addr, String>>>;
+
+const CHANNEL_SIZE: usize = 1_000;
+
+/// Lookup translates between IP addresses and hostnames asynchronously.
+pub trait Lookup: Send {
+    /// Lookup the hostname for the given IP address asynchronously.
+    fn lookup(&mut self, addr: Ipv4Addr);
+}
+
+/// An asynchronous implementation of the Lookup trait.
+pub struct Resolver {
+    tx: Option<Sender<Ipv4Addr>>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Resolver {
+    /// Create a new resolver with the given address translation table.
+    pub fn new(table: IpTable) -> Result<Self> {
+        let runtime = Runtime::new()?;
+        let (tx, mut rx) = mpsc::channel::<Ipv4Addr>(CHANNEL_SIZE);
+        let (resolver, background) = AsyncResolver::from_system_conf()?;
+
+        let handle = Builder::new().name("resolver".into()).spawn({
+            move || {
+                runtime.block_on(async {
+                    let resolver = Arc::new(resolver);
+                    runtime.spawn(background);
+
+                    while let Some(ip) = rx.recv().await {
+                        runtime.spawn({
+                            let resolver = resolver.clone();
+                            let table = table.clone();
+
+                            async move {
+                                let name = resolver
+                                    .reverse_lookup(ip.into())
+                                    .await
+                                    .ok()
+                                    .and_then(|names| names.into_iter().next())
+                                    .map(|name| name.to_string())
+                                    .unwrap_or_else(|| ip.to_string());
```

Comment:
If an IP can't be resolved to a hostname, I'm adding the unresolved IP to the translation table so that we don't try to resolve it again and again. I think this is a change in behavior, let me know if you think it's appropriate.

Reply:
Is this a performance issue? I can understand why it makes sense, I'm just afraid of situations where we cannot reach the DNS server temporarily, for example... what do you think?

Reply:
The idea was to avoid trying to resolve indefinitely an IP address that is not associated with a host name, but you make a very good point, the current implementation would mask transient failures to reach the DNS server. 

After taking a look in more detail to the trust-dns `AsyncResolver` [API](https://docs.rs/trust-dns-resolver/0.18.0-alpha.1/trust_dns_resolver/error/enum.ResolveErrorKind.html), I think we can distinguish between both cases.